### ISSUE TITLE
llm

### ISSUE THREAD

**Issue by BugWriter2** [BugWriter2/banana#49]
# LLAMA

### Background

+ pretraining: the input text is set to max-length -- you can think of a sliding window over a large text corpora
+ finetuning: for bs > 1, we need to pad 

ref: [a comment](https://github.com/huggingface/transformers/issues/22312#issuecomment-1556110284)

### Practice(<=transformers[4.31.0])

finetune with new `[PAD]` token:

ref: [llama recipe](https://github.com/facebookresearch/llama-recipes/blob/03faba661f079ee1ecaeb66deaa6bdec920a7bab/llama_finetuning.py#L124)

caveat: 32k -> 32k+1, thus slower training and inference 

```python
from transformers import LlamaTokenizer, LlamaForCausalLM
tokenizer = LlamaTokenizer.from_pretrained("decapoda-research/llama-7b-hf")
tokenizer.add_special_tokens({'pad_token': '[PAD]'}) #returns 1
model.resize_token_embeddings(len(tokenizer))
```

### Practice(newer version)

use `pad_to_multiple_of` with gpu compute capability >= 7.5

```python
from transformers import LlamaTokenizer, LlamaForCausalLM
tokenizer = LlamaTokenizer.from_pretrained("decapoda-research/llama-7b-hf")
tokenizer.add_special_tokens({'pad_token': '[PAD]'}) #returns 1
model.resize_token_embeddings(len(tokenizer), pad_to_multiple_of=128)
```

ref: [pr](https://github.com/huggingface/transformers/pull/25088/files)

**Comment by BugWriter2** [comment]
# TODO list

## CLM
+ https://medium.com/@tom_21755/understanding-causal-llms-masked-llm-s-and-seq2seq-a-guide-to-language-model-training-d4457bbd07fa
+ https://huggingface.co/learn/nlp-course/chapter7/6
